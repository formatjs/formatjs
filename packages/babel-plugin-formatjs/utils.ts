import * as t from '@babel/types'
import {parse} from '@formatjs/icu-messageformat-parser'
import {interpolateName} from '@formatjs/ts-transformer'

import {
  Options,
  ExtractedMessageDescriptor,
  MessageDescriptor,
  MessageDescriptorPath,
} from './types'
import {NodePath} from '@babel/core'

const DESCRIPTOR_PROPS = new Set<keyof MessageDescriptorPath>([
  'id',
  'description',
  'defaultMessage',
])

function evaluatePath(path: NodePath<any>): string {
  const evaluated = path.evaluate()
  if (evaluated.confident) {
    return evaluated.value
  }

  throw path.buildCodeFrameError(
    '[React Intl] Messages must be statically evaluate-able for extraction.'
  )
}

export function getMessageDescriptorKey(path: NodePath<any>) {
  if (path.isIdentifier() || path.isJSXIdentifier()) {
    return path.node.name
  }

  return evaluatePath(path)
}

function getMessageDescriptorValue(
  path?:
    | NodePath<t.StringLiteral>
    | NodePath<t.JSXExpressionContainer>
    | NodePath<t.TemplateLiteral>,
  isMessageNode?: boolean
) {
  if (!path) {
    return ''
  }
  if (path.isJSXExpressionContainer()) {
    // If this is already compiled, no need to recompiled it
    if (isMessageNode && path.get('expression').isArrayExpression()) {
      return ''
    }
    path = path.get('expression') as NodePath<t.StringLiteral>
  }

  // Always trim the Message Descriptor values.
  const descriptorValue = evaluatePath(path)

  return descriptorValue
}

export function createMessageDescriptor(
  propPaths: [
    NodePath<t.JSXIdentifier> | NodePath<t.Identifier>,
    NodePath<t.StringLiteral> | NodePath<t.JSXExpressionContainer>
  ][]
): MessageDescriptorPath {
  return propPaths.reduce(
    (hash: MessageDescriptorPath, [keyPath, valuePath]) => {
      const key = getMessageDescriptorKey(
        keyPath
      ) as keyof MessageDescriptorPath

      if (DESCRIPTOR_PROPS.has(key)) {
        hash[key] = valuePath
      }

      return hash
    },
    {
      id: undefined,
      defaultMessage: undefined,
      description: undefined,
    }
  )
}

export function evaluateMessageDescriptor(
  descriptorPath: MessageDescriptorPath,
  isJSXSource = false,
  filename: string | undefined,
  idInterpolationPattern?: string,
  overrideIdFn?: Options['overrideIdFn'],
  preserveWhitespace?: Options['preserveWhitespace']
) {
  let id = getMessageDescriptorValue(descriptorPath.id)
  const defaultMessage = getICUMessageValue(
    descriptorPath.defaultMessage,
    {
      isJSXSource,
    },
    preserveWhitespace
  )
  const description = getMessageDescriptorValue(descriptorPath.description)

  if (overrideIdFn) {
    id = overrideIdFn(id, defaultMessage, description, filename)
  } else if (!id && idInterpolationPattern && defaultMessage) {
    id = interpolateName(
      {resourcePath: filename} as any,
      idInterpolationPattern,
      {
        content: description
          ? `${defaultMessage}#${description}`
          : defaultMessage,
      }
    )
  }
  const descriptor: MessageDescriptor = {
    id,
  }

  if (description) {
    descriptor.description = description
  }
  if (defaultMessage) {
    descriptor.defaultMessage = defaultMessage
  }

  return descriptor
}

function getICUMessageValue(
  messagePath?:
    | NodePath<t.StringLiteral>
    | NodePath<t.TemplateLiteral>
    | NodePath<t.JSXExpressionContainer>,
  {isJSXSource = false} = {},
  preserveWhitespace?: Options['preserveWhitespace']
) {
  if (!messagePath) {
    return ''
  }
  let message = getMessageDescriptorValue(messagePath, true)

  if (!preserveWhitespace) {
    message = message.trim().replace(/\s+/gm, ' ')
  }

  try {
    parse(message)
  } catch (parseError) {
    if (
      isJSXSource &&
      messagePath.isLiteral() &&
      message.indexOf('\\\\') >= 0
    ) {
      throw messagePath.buildCodeFrameError(
        '[React Intl] Message failed to parse. ' +
          'It looks like `\\`s were used for escaping, ' +
          "this won't work with JSX string literals. " +
          'Wrap with `{}`. ' +
          'See: http://facebook.github.io/react/docs/jsx-gotchas.html'
      )
    }

    throw messagePath.buildCodeFrameError(
      '[React Intl] Message failed to parse. ' +
        'See: https://formatjs.io/docs/core-concepts/icu-syntax' +
        `\n${parseError}`
    )
  }
  return message
}
const EXTRACTED = Symbol('FormatJSExtracted')
/**
 * Tag a node as extracted
 * Store this in the node itself so that multiple passes work. Specifically
 * if we remove `description` in the 1st pass, 2nd pass will fail since
 * it expect `description` to be there.
 * HACK: We store this in the node instance since this persists across
 * multiple plugin runs
 * @param path
 */
export function tagAsExtracted(path: NodePath<any>) {
  path.node[EXTRACTED] = true
}
/**
 * Check if a node was extracted
 * @param path
 */
export function wasExtracted(path: NodePath<any>) {
  return !!path.node[EXTRACTED]
}

/**
 * Store a message in our global messages
 * @param messageDescriptor
 * @param path
 * @param opts
 * @param filename
 * @param messages
 */
export function storeMessage(
  {id, description, defaultMessage}: MessageDescriptor,
  path: NodePath<any>,
  {extractSourceLocation}: Options,

  filename: string | undefined,
  messages: ExtractedMessageDescriptor[]
) {
  if (!id && !defaultMessage) {
    throw path.buildCodeFrameError(
      '[React Intl] Message Descriptors require an `id` or `defaultMessage`.'
    )
  }

  let loc = {}
  if (extractSourceLocation) {
    loc = {
      file: filename,
      ...path.node.loc,
    }
  }
  messages.push({id, description, defaultMessage, ...loc})
}
