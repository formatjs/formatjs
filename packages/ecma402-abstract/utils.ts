import {memoize, strategies} from '@formatjs/fast-memoize'
/**
 * Cannot do Math.log(x) / Math.log(10) bc if IEEE floating point issue
 * @param x number
 */
export function getMagnitude(x: number): number {
  // Cannot count string length via Number.toString because it may use scientific notation
  // for very small or very large numbers.
  return Math.floor(Math.log(x) * Math.LOG10E)
}

export function repeat(s: string, times: number): string {
  if (typeof s.repeat === 'function') {
    return s.repeat(times)
  }
  const arr = new Array(times)
  for (let i = 0; i < arr.length; i++) {
    arr[i] = s
  }
  return arr.join('')
}

export function setInternalSlot<
  Instance extends object,
  Internal extends object,
  Field extends keyof Internal,
>(
  map: WeakMap<Instance, Internal>,
  pl: Instance,
  field: Field,
  value: NonNullable<Internal>[Field]
) {
  if (!map.get(pl)) {
    map.set(pl, Object.create(null))
  }
  const slots = map.get(pl)!
  slots[field] = value
}

export function setMultiInternalSlots<
  Instance extends object,
  Internal extends object,
  K extends keyof Internal,
>(
  map: WeakMap<Instance, Internal>,
  pl: Instance,
  props: Pick<NonNullable<Internal>, K>
) {
  for (const k of Object.keys(props) as K[]) {
    setInternalSlot(map, pl, k, props[k])
  }
}

export function getInternalSlot<
  Instance extends object,
  Internal extends object,
  Field extends keyof Internal,
>(
  map: WeakMap<Instance, Internal>,
  pl: Instance,
  field: Field
): Internal[Field] {
  return getMultiInternalSlots(map, pl, field)[field]
}

export function getMultiInternalSlots<
  Instance extends object,
  Internal extends object,
  Field extends keyof Internal,
>(
  map: WeakMap<Instance, Internal>,
  pl: Instance,
  ...fields: Field[]
): Pick<Internal, Field> {
  const slots = map.get(pl)
  if (!slots) {
    throw new TypeError(`${pl} InternalSlot has not been initialized`)
  }
  return fields.reduce(
    (all, f) => {
      all[f] = slots[f]
      return all
    },
    Object.create(null) as Pick<Internal, Field>
  )
}

export interface LiteralPart {
  type: 'literal'
  value: string
}

export function isLiteralPart(
  patternPart: LiteralPart | {type: string; value?: string}
): patternPart is LiteralPart {
  return patternPart.type === 'literal'
}

/*
  17 ECMAScript Standard Built-in Objects:
    Every built-in Function object, including constructors, that is not
    identified as an anonymous function has a name property whose value
    is a String.

    Unless otherwise specified, the name property of a built-in Function
    object, if it exists, has the attributes { [[Writable]]: false,
    [[Enumerable]]: false, [[Configurable]]: true }.
*/
export function defineProperty<T extends object>(
  target: T,
  name: string | symbol,
  {value}: {value: any} & ThisType<any>
) {
  Object.defineProperty(target, name, {
    configurable: true,
    enumerable: false,
    writable: true,
    value,
  })
}

/**
 * 7.3.5 CreateDataProperty
 * @param target
 * @param name
 * @param value
 */
export function createDataProperty<T extends object>(
  target: T,
  name: string | symbol,
  value: any
) {
  Object.defineProperty(target, name, {
    configurable: true,
    enumerable: true,
    writable: true,
    value,
  })
}

export const UNICODE_EXTENSION_SEQUENCE_REGEX = /-u(?:-[0-9a-z]{2,8})+/gi

export function invariant(
  condition: boolean,
  message: string,
  Err: any = Error
): asserts condition {
  if (!condition) {
    throw new Err(message)
  }
}

export const createMemoizedNumberFormat: (
  ...args: ConstructorParameters<typeof Intl.NumberFormat>
) => Intl.NumberFormat = memoize(
  (...args: ConstructorParameters<typeof Intl.NumberFormat>) =>
    new Intl.NumberFormat(...args),
  {
    strategy: strategies.variadic,
  }
)

export const createMemoizedDateTimeFormat: (
  ...args: ConstructorParameters<typeof Intl.DateTimeFormat>
) => Intl.DateTimeFormat = memoize(
  (...args: ConstructorParameters<typeof Intl.DateTimeFormat>) =>
    new Intl.DateTimeFormat(...args),
  {
    strategy: strategies.variadic,
  }
)

export const createMemoizedPluralRules: (
  ...args: ConstructorParameters<typeof Intl.PluralRules>
) => Intl.PluralRules = memoize(
  (...args: ConstructorParameters<typeof Intl.PluralRules>) =>
    new Intl.PluralRules(...args),
  {
    strategy: strategies.variadic,
  }
)

export const createMemoizedLocale: (
  ...args: ConstructorParameters<typeof Intl.Locale>
) => Intl.Locale = memoize(
  (...args: ConstructorParameters<typeof Intl.Locale>) =>
    new Intl.Locale(...args),
  {
    strategy: strategies.variadic,
  }
)

export const createMemoizedListFormat: (
  ...args: ConstructorParameters<typeof Intl.ListFormat>
) => Intl.ListFormat = memoize(
  (...args: ConstructorParameters<typeof Intl.ListFormat>) =>
    new Intl.ListFormat(...args),
  {
    strategy: strategies.variadic,
  }
)
