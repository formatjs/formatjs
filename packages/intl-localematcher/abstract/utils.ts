import {data as jsonData} from './languageMatching'
import {regions} from './regions.generated'
export const UNICODE_EXTENSION_SEQUENCE_REGEX: RegExp =
  /-u(?:-[0-9a-z]{2,8})+/gi

export function invariant(
  condition: boolean,
  message: string,
  Err: any = Error
): asserts condition {
  if (!condition) {
    throw new Err(message)
  }
}

// This is effectively 2 languages in 2 different regions in the same cluster
const DEFAULT_MATCHING_THRESHOLD = 838

interface LSR {
  language: string
  script: string
  region: string
}

interface LanguageMatchInfo {
  supported: string
  desired: string
  distance: number
  oneway: boolean
}

interface LanguageInfo {
  matches: LanguageMatchInfo[]
  matchVariables: Record<string, string[]>
  paradigmLocales: string[]
}

let PROCESSED_DATA: LanguageInfo | undefined

function processData(): LanguageInfo {
  if (!PROCESSED_DATA) {
    const paradigmLocales = jsonData.supplemental.languageMatching[
      'written-new'
    ][0]?.paradigmLocales?._locales.split(' ') as any
    const matchVariables = jsonData.supplemental.languageMatching[
      'written-new'
    ].slice(1, 5) as any[]
    const data = jsonData.supplemental.languageMatching['written-new'].slice(5)
    const matches = data.map(d => {
      const key = Object.keys(d)[0] as string
      const value = (
        d as Record<
          string,
          {
            _desired: string
            _distance: string
            oneway?: string
          }
        >
      )[key]
      return {
        supported: key,
        desired: value._desired,
        distance: +value._distance,
        oneway: value.oneway === 'true' ? true : false,
      }
    }, {})
    PROCESSED_DATA = {
      matches,
      matchVariables: matchVariables.reduce<Record<string, string[]>>(
        (all, d) => {
          const key = Object.keys(d)[0] as string
          const value = d[key]
          all[key.slice(1)] = value._value.split('+')
          return all
        },
        {}
      ),
      paradigmLocales: [
        ...paradigmLocales,
        ...paradigmLocales.map((l: string) =>
          new Intl.Locale(l.replace(/_/g, '-')).maximize().toString()
        ),
      ],
    }
  }

  return PROCESSED_DATA
}

function isMatched(
  locale: LSR,
  languageMatchInfoLocale: string,
  matchVariables: Record<string, string[]>
): boolean {
  const [language, script, region] = languageMatchInfoLocale.split('-')
  let matches = true
  if (region && region[0] === '$') {
    const shouldInclude = region[1] !== '!'
    const matchRegions = shouldInclude
      ? matchVariables[region.slice(1)]
      : matchVariables[region.slice(2)]
    const expandedMatchedRegions = matchRegions
      .map(r => regions[r] || [r])
      .reduce((all, list) => [...all, ...list], [])
    matches &&= !(
      expandedMatchedRegions.indexOf(locale.region || '') > -1 !=
      shouldInclude
    )
  } else {
    matches &&= locale.region
      ? region === '*' || region === locale.region
      : true
  }
  matches &&= locale.script ? script === '*' || script === locale.script : true
  matches &&= locale.language
    ? language === '*' || language === locale.language
    : true
  return matches
}

function serializeLSR(lsr: LSR): string {
  return [lsr.language, lsr.script, lsr.region].filter(Boolean).join('-')
}

function findMatchingDistanceForLSR(
  desired: LSR,
  supported: LSR,
  data: LanguageInfo
): number {
  for (const d of data.matches) {
    let matches =
      isMatched(desired, d.desired, data.matchVariables) &&
      isMatched(supported, d.supported, data.matchVariables)
    if (!d.oneway && !matches) {
      matches =
        isMatched(desired, d.supported, data.matchVariables) &&
        isMatched(supported, d.desired, data.matchVariables)
    }
    if (matches) {
      const distance = d.distance * 10
      if (
        data.paradigmLocales.indexOf(serializeLSR(desired)) > -1 !=
        data.paradigmLocales.indexOf(serializeLSR(supported)) > -1
      ) {
        return distance - 1
      }
      return distance
    }
  }
  throw new Error('No matching distance found')
}

export function findMatchingDistance(
  desired: string,
  supported: string
): number {
  const desiredLocale = new Intl.Locale(desired).maximize()
  const supportedLocale = new Intl.Locale(supported).maximize()
  const desiredLSR: LSR = {
    language: desiredLocale.language,
    script: desiredLocale.script || '',
    region: desiredLocale.region || '',
  }
  const supportedLSR: LSR = {
    language: supportedLocale.language,
    script: supportedLocale.script || '',
    region: supportedLocale.region || '',
  }
  let matchingDistance = 0

  const data = processData()

  if (desiredLSR.language !== supportedLSR.language) {
    matchingDistance += findMatchingDistanceForLSR(
      {
        language: desiredLocale.language,
        script: '',
        region: '',
      },
      {
        language: supportedLocale.language,
        script: '',
        region: '',
      },
      data
    )
  }

  if (desiredLSR.script !== supportedLSR.script) {
    matchingDistance += findMatchingDistanceForLSR(
      {
        language: desiredLocale.language,
        script: desiredLSR.script,
        region: '',
      },
      {
        language: supportedLocale.language,
        script: supportedLSR.script,
        region: '',
      },
      data
    )
  }

  if (desiredLSR.region !== supportedLSR.region) {
    matchingDistance += findMatchingDistanceForLSR(
      desiredLSR,
      supportedLSR,
      data
    )
  }

  return matchingDistance
}

interface LocaleMatchingResult {
  distances: Record<string, Record<string, number>>
  matchedSupportedLocale?: string
  matchedDesiredLocale?: string
}

export function findBestMatch(
  requestedLocales: readonly string[],
  supportedLocales: readonly string[],
  threshold: number = DEFAULT_MATCHING_THRESHOLD
): LocaleMatchingResult {
  let lowestDistance = Infinity
  let result: LocaleMatchingResult = {
    matchedDesiredLocale: '',
    distances: {},
  }
  requestedLocales.forEach((desired, i) => {
    if (!result.distances[desired]) {
      result.distances[desired] = {}
    }
    supportedLocales.forEach(supported => {
      // Add some weight to the distance based on the order of the supported locales
      // Add penalty for the order of the requested locales, which currently is 0 since ECMA-402
      // doesn't really have room for weighted locales like `en; q=0.1`
      const distance = findMatchingDistance(desired, supported) + 0 + i * 40

      result.distances[desired][supported] = distance
      if (distance < lowestDistance) {
        lowestDistance = distance
        result.matchedDesiredLocale = desired
        result.matchedSupportedLocale = supported
      }
    })
  })

  if (lowestDistance >= threshold) {
    result.matchedDesiredLocale = undefined
    result.matchedSupportedLocale = undefined
  }

  return result
}
