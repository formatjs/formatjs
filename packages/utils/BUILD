load("@aspect_bazel_lib//lib:copy_to_bin.bzl", "copy_to_bin")
load("@aspect_bazel_lib//lib:write_source_files.bzl", "write_source_files")
load("@aspect_rules_js//npm:defs.bzl", "npm_package")
load("//tools:index.bzl", "generate_src_file", "package_json_test", "ts_compile")
load("//tools:jest.bzl", "jest_test")

exports_files([
    "package.json",
    "tsconfig.json",
])

PACKAGE_NAME = "utils"

npm_package(
    name ="pkg",
    srcs = [
        "LICENSE.md",
        "README.md",
        "package.json",
        ":dist",
        ":dist-esm",
    ],
    package = "@formatjs/%s" % PACKAGE_NAME,
    visibility = ["//visibility:public"],
)

SRCS = glob([
    "src/**/*",
    "*.ts",
])

SRC_DEPS = [
]

ts_compile(
    name = "dist",
    srcs = [":srcs"],
    skip_esm = False,
    deps = SRC_DEPS,
)

write_source_files(
    name = "tsconfig_json",
    files = {"tsconfig.json": "//tools:tsconfig.golden.json"},
)

package_json_test(
    name = "package_json_test",
    deps = SRC_DEPS,
)

jest_test(
    name = "unit_test",
    data = [":srcs"] + SRC_DEPS + glob([
        "tests/*.ts*",
        "tests/fixtures/*.ts*",
    ]),
    snapshots = glob(["tests/__snapshots__/*.snap"]),
)

generate_src_file(
    name = "default-currency",
    src = "src/defaultCurrencyData.generated.json",
    data = [
        "//:node_modules/cldr-core",
    ],
    entry_point = "tools/extract-default-currency.ts",
)

copy_to_bin(name = "srcs", srcs=SRCS)
