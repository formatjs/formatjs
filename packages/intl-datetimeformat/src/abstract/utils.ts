import {IntlDateTimeFormatInternal, TABLE_6} from '@formatjs/ecma402-abstract'

export const DATE_TIME_PROPS: Array<
  keyof Pick<IntlDateTimeFormatInternal, TABLE_6>
> = [
  'weekday',
  'era',
  'year',
  'month',
  'day',
  'dayPeriod',
  'hour',
  'minute',
  'second',
  'fractionalSecondDigits',
  'timeZoneName',
]

export const removalPenalty = 120
export const additionPenalty = 20
export const differentNumericTypePenalty = 15
export const longLessPenalty = 8
export const longMorePenalty = 6
export const shortLessPenalty = 6
export const shortMorePenalty = 3
export const offsetPenalty = 1
