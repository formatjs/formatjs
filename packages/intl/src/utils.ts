import {
  IntlCache,
  CustomFormats,
  Formatters,
  OnErrorFn,
  ResolvedIntlConfig,
} from './types'
import {IntlMessageFormat} from 'intl-messageformat'
import memoize, {Cache, strategies} from '@formatjs/fast-memoize'
import {UnsupportedFormatterError} from './error'
import {DateTimeFormat} from '@formatjs/ecma402-abstract'

export function filterProps<T extends Record<string, any>, K extends string>(
  props: T,
  whitelist: Array<K>,
  defaults: Partial<T> = {}
): Pick<T, K> {
  return whitelist.reduce((filtered, name) => {
    if (name in props) {
      filtered[name] = props[name]
    } else if (name in defaults) {
      filtered[name] = defaults[name]!
    }

    return filtered
  }, {} as Pick<T, K>)
}

const defaultErrorHandler: OnErrorFn = error => {
  // @ts-ignore just so we don't need to declare dep on @types/node
  if (process.env.NODE_ENV !== 'production') {
    console.error(error)
  }
}

export const DEFAULT_INTL_CONFIG: Pick<
  ResolvedIntlConfig<any>,
  | 'fallbackOnEmptyString'
  | 'formats'
  | 'messages'
  | 'timeZone'
  | 'defaultLocale'
  | 'defaultFormats'
  | 'onError'
> = {
  formats: {},
  messages: {},
  timeZone: undefined,

  defaultLocale: 'en',
  defaultFormats: {},

  fallbackOnEmptyString: true,

  onError: defaultErrorHandler,
}

export function createIntlCache(): IntlCache {
  return {
    dateTime: {},
    number: {},
    message: {},
    relativeTime: {},
    pluralRules: {},
    list: {},
    displayNames: {},
  }
}

function createFastMemoizeCache<V>(store: Record<string, V>): Cache<string, V> {
  return {
    create() {
      return {
        get(key) {
          return store[key]
        },
        set(key, value) {
          store[key] = value
        },
      }
    },
  }
}

/**
 * Create intl formatters and populate cache
 * @param cache explicit cache to prevent leaking memory
 */
export function createFormatters(
  cache: IntlCache = createIntlCache()
): Formatters {
  const RelativeTimeFormat = (Intl as any).RelativeTimeFormat
  const ListFormat = (Intl as any).ListFormat
  const DisplayNames = (Intl as any).DisplayNames
  const getDateTimeFormat = memoize(
    (...args) => new Intl.DateTimeFormat(...args) as DateTimeFormat,
    {
      cache: createFastMemoizeCache(cache.dateTime),
      strategy: strategies.variadic,
    }
  )
  const getNumberFormat = memoize((...args) => new Intl.NumberFormat(...args), {
    cache: createFastMemoizeCache(cache.number),
    strategy: strategies.variadic,
  })
  const getPluralRules = memoize((...args) => new Intl.PluralRules(...args), {
    cache: createFastMemoizeCache(cache.pluralRules),
    strategy: strategies.variadic,
  })
  return {
    getDateTimeFormat,
    getNumberFormat,
    getMessageFormat: memoize(
      (message, locales, overrideFormats, opts) =>
        new IntlMessageFormat(message, locales, overrideFormats, {
          formatters: {
            getNumberFormat,
            getDateTimeFormat,
            getPluralRules,
          },
          ...(opts || {}),
        }),
      {
        cache: createFastMemoizeCache(cache.message),
        strategy: strategies.variadic,
      }
    ),
    getRelativeTimeFormat: memoize(
      (...args) => new RelativeTimeFormat(...args),
      {
        cache: createFastMemoizeCache(cache.relativeTime),
        strategy: strategies.variadic,
      }
    ),
    getPluralRules,
    getListFormat: memoize((...args) => new ListFormat(...args), {
      cache: createFastMemoizeCache(cache.list),
      strategy: strategies.variadic,
    }),
    getDisplayNames: memoize((...args) => new DisplayNames(...args), {
      cache: createFastMemoizeCache(cache.displayNames),
      strategy: strategies.variadic,
    }),
  }
}

export function getNamedFormat<T extends keyof CustomFormats>(
  formats: CustomFormats,
  type: T,
  name: string,
  onError: OnErrorFn
):
  | Intl.NumberFormatOptions
  | Intl.DateTimeFormatOptions
  | Intl.RelativeTimeFormatOptions
  | undefined {
  const formatType = formats && formats[type]
  let format
  if (formatType) {
    format = formatType[name]
  }
  if (format) {
    return format
  }

  onError(new UnsupportedFormatterError(`No ${type} format named: ${name}`))
}
