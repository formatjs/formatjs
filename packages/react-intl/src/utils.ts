import {ResolvedIntlConfig} from './types'
import * as React from 'react'
import {FormatXMLElementFn} from 'intl-messageformat'
import {invariant} from '@formatjs/ecma402-abstract'

import {DEFAULT_INTL_CONFIG as CORE_DEFAULT_INTL_CONFIG} from '@formatjs/intl'

export function invariantIntlContext(intl?: any): asserts intl {
  invariant(
    intl,
    '[React Intl] Could not find required `intl` object. ' +
      '<IntlProvider> needs to exist in the component ancestry.'
  )
}

export const DEFAULT_INTL_CONFIG: Pick<
  ResolvedIntlConfig,
  | 'fallbackOnEmptyString'
  | 'formats'
  | 'messages'
  | 'timeZone'
  | 'textComponent'
  | 'defaultLocale'
  | 'defaultFormats'
  | 'onError'
> = {
  ...CORE_DEFAULT_INTL_CONFIG,
  textComponent: React.Fragment,
}

/**
 * Takes a `formatXMLElementFn`, and composes it in function, which passes
 * argument `parts` through, assigning unique key to each part, to prevent
 * "Each child in a list should have a unique "key"" React error.
 * @param formatXMLElementFn
 */
export function assignUniqueKeysToParts(
  formatXMLElementFn: FormatXMLElementFn<React.ReactNode>
): FormatXMLElementFn<React.ReactNode> {
  return function (parts: any) {
    // eslint-disable-next-line prefer-rest-params
    return formatXMLElementFn(React.Children.toArray(parts)) as any
  }
}

export function shallowEqual<
  T extends Record<string, unknown> = Record<string, unknown>
>(objA?: T, objB?: T) {
  if (objA === objB) {
    return true
  }

  if (!objA || !objB) {
    return false
  }

  var aKeys = Object.keys(objA)
  var bKeys = Object.keys(objB)
  var len = aKeys.length

  if (bKeys.length !== len) {
    return false
  }

  for (var i = 0; i < len; i++) {
    var key = aKeys[i]

    if (
      objA[key] !== objB[key] ||
      !Object.prototype.hasOwnProperty.call(objB, key)
    ) {
      return false
    }
  }

  return true
}
