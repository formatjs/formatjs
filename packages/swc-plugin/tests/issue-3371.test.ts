import {join} from 'path'
import {compile, FIXTURES_DIR} from './utils'
import {basename} from 'path'

test('GH issue #3371', async function () {
  const output = await compile(join(FIXTURES_DIR, `resourcePath.tsx`), {
    ast: true,
    overrideIdFn: (id, defaultMessage, description, filePath) => {
      return `HELLO.${id}.${
        defaultMessage!.length
      }.${typeof description}.${basename(filePath || '')}`
    },
  })
  expect(output.msgs).toMatchInlineSnapshot(`
    [
      {
        "defaultMessage": "props {intl}",
        "description": "bar",
        "id": "HELLO..12.string.resourcePath.tsx",
      },
      {
        "defaultMessage": "this props {intl}",
        "description": "bar",
        "id": "HELLO..17.string.resourcePath.tsx",
      },
      {
        "defaultMessage": "foo {bar}",
        "description": "bar",
        "id": "HELLO..9.string.resourcePath.tsx",
      },
    ]
  `)
  expect(output.code).toMatchSnapshot()
})
