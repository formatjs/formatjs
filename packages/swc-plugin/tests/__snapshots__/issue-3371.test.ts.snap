// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GH issue #3371 2`] = `
"function _defineProperty(obj, key, value) {
    if (key in obj) {
        Object.defineProperty(obj, key, {
            value: value,
            enumerable: true,
            configurable: true,
            writable: true
        });
    } else {
        obj[key] = value;
    }
    return obj;
}
var _obj, _obj1, _obj2;
export function foo() {
    props.intl.formatMessage((_obj = {
        id: \\"HELLO..12.string.resourcePath.tsx\\",
        defaultMessage: [
            {
                type: 0,
                value: \\"props \\"
            },
            {
                type: 1,
                value: \\"intl\\"
            }, 
        ]
    }, _defineProperty(_obj, \\"defaultMessage\\", 'props {intl}'), _defineProperty(_obj, \\"description\\", 'bar'), _obj), {
        bar: 'bar'
    });
    this.props.intl.formatMessage((_obj1 = {
        id: \\"HELLO..17.string.resourcePath.tsx\\",
        defaultMessage: [
            {
                type: 0,
                value: \\"this props \\"
            },
            {
                type: 1,
                value: \\"intl\\"
            }, 
        ]
    }, _defineProperty(_obj1, \\"defaultMessage\\", 'this props {intl}'), _defineProperty(_obj1, \\"description\\", 'bar'), _obj1), {
        bar: 'bar'
    });
    return intl.formatMessage((_obj2 = {
        id: \\"HELLO..9.string.resourcePath.tsx\\",
        defaultMessage: [
            {
                type: 0,
                value: \\"foo \\"
            },
            {
                type: 1,
                value: \\"bar\\"
            }, 
        ]
    }, _defineProperty(_obj2, \\"defaultMessage\\", 'foo {bar}'), _defineProperty(_obj2, \\"description\\", 'bar'), _obj2), {
        bar: 'bar'
    });
}
"
`;
